source/extensions/filters/common/rbac/matchers.cc

```diff
@@ -0,0 +1,140 @@
+  case envoy::config::rbac::v2alpha::Principal::IdentifierCase::kSourceIp:
+    return std::make_shared<const IPMatcher>(principal.source_ip(), false);
+  case envoy::config::rbac::v2alpha::Principal::IdentifierCase::kHeader:
+    return std::make_shared<const HeaderMatcher>(principal.header());
+  case envoy::config::rbac::v2alpha::Principal::IdentifierCase::kAny:
+    return std::make_shared<const AlwaysMatcher>();
+  default:
+    NOT_REACHED;
+  }
+}
+
+AndMatcher::AndMatcher(const envoy::config::rbac::v2alpha::Permission_Set& set) {
+  for (const auto& rule : set.rules()) {
+    matchers_.push_back(Matcher::create(rule));
+  }
+}
+
+AndMatcher::AndMatcher(const envoy::config::rbac::v2alpha::Principal_Set& set) {
+  for (const auto& id : set.ids()) {
+    matchers_.push_back(Matcher::create(id));
+  }
+}
+
+bool AndMatcher::matches(const Network::Connection& connection,
+                         const Envoy::Http::HeaderMap& headers) const {
+  for (const auto& matcher : matchers_) {
+    if (!matcher->matches(connection, headers)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+OrMatcher::OrMatcher(
+    const Protobuf::RepeatedPtrField<::envoy::config::rbac::v2alpha::Permission>& rules) {
+  for (const auto& rule : rules) {
+    matchers_.push_back(Matcher::create(rule));
+  }
+}
+
+OrMatcher::OrMatcher(
+    const Protobuf::RepeatedPtrField<::envoy::config::rbac::v2alpha::Principal>& ids) {
+  for (const auto& id : ids) {
+    matchers_.push_back(Matcher::create(id));
+  }
+}
+
+bool OrMatcher::matches(const Network::Connection& connection,
+                        const Envoy::Http::HeaderMap& headers) const {
+  for (const auto& matcher : matchers_) {
+    if (matcher->matches(connection, headers)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool HeaderMatcher::matches(const Network::Connection&,
+                            const Envoy::Http::HeaderMap& headers) const {
+  return Envoy::Http::HeaderUtility::matchHeaders(headers, header_);
+}
+
+bool IPMatcher::matches(const Network::Connection& connection,
+                        const Envoy::Http::HeaderMap&) const {
+  const Envoy::Network::Address::InstanceConstSharedPtr& ip =
+      destination_ ? connection.localAddress() : connection.remoteAddress();
+
+  return range_.isInRange(*ip.get());
+}
+
+bool PortMatcher::matches(const Network::Connection& connection,
+                          const Envoy::Http::HeaderMap&) const {
+  const Envoy::Network::Address::Ip* ip = connection.localAddress().get()->ip();
+  return ip && ip->port() == port_;
+}
+
+bool AuthenticatedMatcher::matches(const Network::Connection& connection,
+                                   const Envoy::Http::HeaderMap&) const {
+  const auto* ssl = connection.ssl();
+  if (!ssl) { // connection was not authenticated
+    return false;
+  } else if (name_.empty()) { // matcher allows any subject
+    return true;
+  }
+
+  std::string principal = ssl->uriSanPeerCertificate();
```

Comment:
For Istio, my understanding is local RBAC filter is going to chain with authN filter, which already fished out identiy for both mtls/JWT cases and set principle in request header; so here may need to use principle from header directly. 

https://github.com/istio/proxy/blob/master/src/envoy/http/authn/filter_context.cc#L56 
https://github.com/istio/proxy/blob/master/src/envoy/http/authn/http_filter.cc#L79